v2/header.go

```diff
@@ -117,3 +123,40 @@ func XGoogHeader(keyval ...string) string {
 	}
 	return buf.String()[1:]
 }
+
+// InsertMetadataIntoOutgoingContext is for use by the Google Cloud Libraries
+// only.
+//
+// InsertMetadataIntoOutgoingContext returns a new context that merges the
+// provided keyvals metadata pairs with any existing metadata/headers in the
+// provided context. keyvals should have a corresponding value for every key
+// provided. If there is an odd number of keyvals this method will panic.
+func InsertMetadataIntoOutgoingContext(ctx context.Context, keyvals ...string) context.Context {
+	return metadata.NewOutgoingContext(ctx, insertMetadata(ctx, keyvals...))
+}
+
+// BuildHeaders is for use by the Google Cloud Libraries only.
+//
+// BuildHeaders returns a new http.Header that merges the provided
+// keyvals header pairs with any existing metadata/headers in the provided
+// context. keyvals should have a corresponding value for every key provided.
+// If there is an odd number of keyvals this method will panic.
+func BuildHeaders(ctx context.Context, keyvals ...string) http.Header {
+	return http.Header(insertMetadata(ctx, keyvals...))
+}
+
+func insertMetadata(ctx context.Context, keyvals ...string) metadata.MD {
+	if len(keyvals)%2 != 0 {
+		panic(fmt.Sprintf("gax: an even number of key value pairs must be provided, got %d", len(keyvals)))
+	}
+	out, _ := metadata.FromOutgoingContext(ctx)
+	out = out.Copy()
+	headers := callctx.HeadersFromContext(ctx)
```

Comment:
Since these are headers potentially coming from users I wonder if they should happen last? Also I think it is fine that we are appending to the headers here, but maybe we should explicitly document this design choice. As is a user can't overwrite an existing key, which I think is fine? But it should at least be stated

Reply:
The existing func created by the generator uses the user-provided metadata in the context as the base, then adds the client-provided arguments. I thought I should therefore add the user-provided headers in the context before the client-provided arguments. Let me know how I should order everything if you think this is incorrect. Happy to make the change.

Reply:
Updated docs to note that existing keys are not overwritten, but values are instead appended.